// @kinvolk/headlamp-plugin/lib/components/common/Tooltip/TooltipLight.js
import { jsx as _jsx } from "react/jsx-runtime";
import Tooltip from '@mui/material/Tooltip';
export default function TooltipLight(props) {
    const { children, interactive = true, ...rest } = props;
    const disableInteractive = !interactive;
    if (typeof children === 'string') {
        return (_jsx(Tooltip, { disableInteractive: disableInteractive, sx: theme => ({
                backgroundColor: theme.palette.background.default,
                color: theme.palette.resourceToolTip.color,
                boxShadow: theme.shadows[1],
                fontSize: '1rem',
                whiteSpace: 'pre-line',
            }), ...rest, children: _jsx("span", { children: children }) }));
    }
    return (_jsx(Tooltip, { ...props, 
        // children prop in the mui Tooltip is defined as ReactElement which is not totally correct
        // string should be a valid child and is used a lot in this project
        // but it's not included in the ReactElement type
        children: props.children }));
}
