// @kinvolk/headlamp-plugin/lib/components/common/SectionHeader.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box } from '@mui/material';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
export default function SectionHeader(props) {
    const { noPadding = false, headerStyle = 'main', titleSideActions = [] } = props;
    const actions = props.actions || [];
    const titleVariants = {
        main: 'h1',
        subsection: 'h2',
        normal: 'h3',
        label: 'h4',
    };
    return (_jsxs(Grid, { container: true, alignItems: "center", justifyContent: "space-between", sx: theme => ({
            padding: theme.spacing(noPadding ? 0 : 2),
            paddingTop: theme.spacing(noPadding ? 0 : 3),
            paddingRight: '0',
        }), spacing: 2, children: [_jsx(Grid, { item: true, children: (!!props.title || titleSideActions.length > 0) && (_jsxs(Box, { display: "flex", alignItems: "center", children: [!!props.title && (_jsx(Typography, { variant: titleVariants[headerStyle], noWrap: true, sx: theme => ({
                                ...theme.palette.headerStyle[headerStyle || 'normal'],
                                whiteSpace: 'pre-wrap',
                            }), children: props.title })), !!titleSideActions && (_jsx(Box, { ml: 1, justifyContent: "center", children: titleSideActions }))] })) }), actions.length > 0 && (_jsx(Grid, { item: true, children: _jsx(Grid, { item: true, container: true, alignItems: "center", justifyContent: "flex-end", children: actions.map((action, i) => (_jsx(Grid, { item: true, children: action }, i))) }) }))] }));
}
