// @kinvolk/headlamp-plugin/lib/components/App/Home/SquareButton.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Icon } from '@iconify/react';
import { useTheme } from '@mui/material';
import ButtonBase from '@mui/material/ButtonBase';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Typography from '@mui/material/Typography';
export default function SquareButton(props) {
    const { icon, iconSize = 50, iconColor, label, primary = false, ...otherProps } = props;
    const theme = useTheme();
    return (_jsx(ButtonBase, { focusRipple: true, ...otherProps, children: _jsx(Card, { sx: {
                width: 140,
                height: 140,
                paddingTop: '24px',
                backgroundColor: primary ? 'text.primary' : 'squareButton.background',
            }, children: _jsxs(CardContent, { sx: { textAlign: 'center', paddingTop: 0 }, children: [_jsx(Icon, { icon: icon, width: iconSize, height: iconSize, color: iconColor ||
                            (primary
                                ? theme.palette.getContrastText(theme.palette.text.primary)
                                : theme.palette.getContrastText(theme.palette.squareButton.background)) }), _jsx(Typography, { color: "textSecondary", gutterBottom: true, title: label, sx: {
                            textOverflow: 'ellipsis',
                            whiteSpace: 'nowrap',
                            overflow: 'hidden',
                            display: 'block',
                            fontSize: '1rem',
                            paddingTop: '8px',
                            color: primary
                                ? theme.palette.getContrastText(theme.palette.text.primary)
                                : theme.palette.getContrastText(theme.palette.squareButton.background),
                        }, children: label })] }) }) }));
}
