// @kinvolk/headlamp-plugin/lib/components/resourceMap/details/GraphNodeDetails.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Card } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { ActionButton } from '../../common';
import { KubeObjectDetails } from './KubeNodeDetails';
const kubeNodeDetailsSection = {
    id: 'kubeObjectDetails',
    nodeType: 'kubeObject',
    render: node => (_jsx(KubeObjectDetails, { resource: node.data.resource }, node.id)),
};
const defaultSections = [kubeNodeDetailsSection];
/**
 * Side panel display information about a selected Node
 */
export function GraphNodeDetails({ sections = defaultSections, node, close, }) {
    const { t } = useTranslation();
    return (_jsxs(Card, { sx: theme => ({
            margin: '0',
            padding: '1rem',
            width: '900px',
            overflowY: 'auto',
            flexShrink: 0,
            [theme.breakpoints.down('xl')]: {
                width: '720px',
            },
            [theme.breakpoints.down('lg')]: {
                zIndex: 1,
                position: 'absolute',
                width: '100%',
                minWidth: '100%',
            },
        }), children: [_jsx(Box, { textAlign: "right", children: _jsx(ActionButton, { onClick: () => {
                        close();
                    }, icon: "mdi:close", description: t('Close') }) }), sections.filter(it => it.nodeType === node.type).map(it => it.render(node))] }));
}
