// @kinvolk/headlamp-plugin/lib/lib/k8s/apiProxy/index.js
/*
 * This module was originally taken from the K8dash project before modifications.
 *
 * K8dash is licensed under Apache License 2.0.
 *
 * Copyright © 2020 Eric Herbrandson
 * Copyright © 2020 Kinvolk GmbH
 */
// Basic cluster API functions
export { clusterRequest, patch, post, put, remove, request, } from '../api/v1/clusterRequests';
// Streaming API functions
export { stream, streamResult, streamResults, streamResultsForCluster, } from '../api/v1/streamingApi';
// API factory functions
export { apiFactory, apiFactoryWithNamespace, } from '../api/v1/factories';
// Port forward functions
export { listPortForward, startPortForward, stopOrDeletePortForward } from '../api/v1/portForward';
export { deleteCluster, setCluster, testAuth, testClusterHealth, parseKubeConfig, renameCluster, } from '../api/v1/clusterApi';
export { metrics } from '../api/v1/metricsApi';
export { deletePlugin } from '../api/v1/pluginsApi';
export { drainNodeStatus, drainNode } from '../api/v1/drainNode';
export { apply } from '../api/v1/apply';
