// @kinvolk/headlamp-plugin/lib/components/common/TileChart/TileChart.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import '../../../i18n/config';
import { Box, Paper, Typography } from '@mui/material';
import { PercentageCircle } from '../Chart';
import { TooltipIcon } from '../Tooltip';
export function TileChart(props) {
    const { title, infoTooltip = '', legend, total, data, ...others } = props;
    return (_jsx(Paper, { sx: theme => ({
            background: theme.palette.squareButton.background,
            padding: theme.spacing(2),
            height: '100%',
            maxWidth: '300px',
            margin: '0 auto',
        }), children: _jsxs(Box, { display: "flex", sx: theme => ({
                [theme.breakpoints.down('sm')]: {
                    flexWrap: 'wrap',
                },
            }), children: [_jsxs(Box, { flexGrow: 1, width: "100%", children: [_jsxs(Box, { children: [_jsx(Typography, { sx: theme => ({
                                        fontSize: theme.typography.pxToRem(16),
                                        display: 'inline',
                                        fontWeight: 600,
                                    }), gutterBottom: true, children: title || '' }), infoTooltip && _jsx(TooltipIcon, { children: infoTooltip })] }), _jsx(Typography, { sx: theme => ({
                                fontSize: theme.typography.pxToRem(16),
                                display: 'inline',
                                fontWeight: 400,
                            }), gutterBottom: true, children: legend || '' })] }), _jsx(Box, { children: !!data && (_jsx(PercentageCircle, { data: data, total: total, size: 140, thickness: 11, ...others })) })] }) }));
}
export default TileChart;
