// @kinvolk/headlamp-plugin/lib/components/statefulset/List.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useTranslation } from 'react-i18next';
import StatefulSet from '../../lib/k8s/statefulSet';
import { LightTooltip } from '../common';
import ResourceListView from '../common/Resource/ResourceListView';
export default function StatefulSetList() {
    const { t } = useTranslation('glossary');
    function renderPods(statefulSet) {
        const { readyReplicas, replicas } = statefulSet.status;
        return `${readyReplicas || 0}/${replicas || 0}`;
    }
    return (_jsx(ResourceListView, { title: t('Stateful Sets'), resourceClass: StatefulSet, columns: [
            'name',
            'namespace',
            'cluster',
            {
                id: 'pods',
                label: t('Pods'),
                getValue: statefulSet => renderPods(statefulSet),
                gridTemplate: 0.6,
            },
            {
                id: 'replicas',
                label: t('Replicas'),
                getValue: statefulSet => statefulSet.spec.replicas,
                gridTemplate: 0.6,
            },
            {
                id: 'containers',
                label: t('Containers'),
                getValue: statefulSet => statefulSet
                    .getContainers()
                    .map(c => c.name)
                    .join(', '),
                render: statefulSet => {
                    const containerNames = statefulSet.getContainers().map((c) => c.name);
                    const containerTooltip = containerNames.join('\n');
                    const containerText = containerNames.join(', ');
                    return (_jsx(LightTooltip, { title: containerTooltip, interactive: true, children: containerText }));
                },
            },
            {
                id: 'images',
                label: t('Images'),
                getValue: statefulSet => statefulSet
                    .getContainers()
                    .map(it => it.image)
                    .join(', '),
                render: statefulSet => {
                    const containerImages = statefulSet.getContainers().map((c) => c.image);
                    const containerTooltip = containerImages.join('\n');
                    const containerText = containerImages.join(', ');
                    return (_jsx(LightTooltip, { title: containerTooltip, interactive: true, children: containerText }));
                },
            },
            'age',
        ] }));
}
